### PR TITLE
F/kv parser refinements

### PR DESCRIPTION
**Pull request by bazsi**
This branch merges the two parallel implementations of KVScanner introduced by an earlier pull request. It also adds pair-separator() as an option and does a number of refactors.

Please review and merge.
@csabamajor

### CODE REVIEW

**Comment by bazsi** [comment]
@MrAnno thanks for the fix. So as it seems only the review is needed and this is good to go.

I have split the patches to make that easier, but still its a challenge. let me know if you have questions.

**Comment by bazsi** [comment]
I have folded in the fixup by @MrAnno and also added two minor changes to address review notes that we had IRL.

This should be as good as it gets.

File: lib/str-repr/encode.c [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+str_repr_encode_append(GString *escaped_string, const gchar *str, gssize str_len, const gchar *forbidden_chars)
+{
+  const gchar *apostrophe, *quote;
+
+  if (str_len < 0)
+    str_len = strlen(str);
+
+  if (str_len == 0)
+    {
+      g_string_append_len(escaped_string, "''", 2);
```
Comment (lbudai) [inline lib/str-repr/encode.c]: Shouldn't we return `"\"\""`?

File: lib/str-repr/decode.c [hunk 2]
```diff
@@ -0,0 +1,212 @@
...
+            {
+              cur = new_cur;
+              goto finish;
+            }
+          else
+            {
+              g_string_append_c(value, *cur);
+            }
+          break;
+        }
```
Comment (lbudai) [inline lib/str-repr/decode.c]: Could you extract case branches into separate functions?

File: lib/str-repr/tests/test_encode.c [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+ * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
+ *
+ * As an additional exemption you are allowed to compile & link against the
+ * OpenSSL libraries as published by the OpenSSL project. See the file
+ * COPYING for details.
+ *
+ */
+#include "str-repr/encode.h"
+#include "testutils.h"
+
```
Comment (lbudai) [inline lib/str-repr/tests/test_encode.c]: please use criterion.

File: lib/str-repr/encode.c (hunk 1 above)
Comment (bazsi) [inline lib/str-repr/encode.c]: You mean to use quotes instead of apostrophes? Both should be fine.

File: lib/str-repr/decode.c (hunk 2 above)
Comment (bazsi) [inline lib/str-repr/decode.c]: You mean all branches into separate functions?

File: lib/str-repr/tests/test_decode.c [hunk 4]
```diff
@@ -0,0 +1,165 @@
...
+ * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
+ *
+ * As an additional exemption you are allowed to compile & link against the
+ * OpenSSL libraries as published by the OpenSSL project. See the file
+ * COPYING for details.
+ *
+ */
+#include "str-repr/decode.h"
+#include "testutils.h"
+
```
Comment (bazsi) [inline lib/str-repr/tests/test_decode.c]: Well, this test predates criterion, and I did convert one complicated testcase in this branch.

Yeah, some of this code is that old. Could this go in with these tests please? Pretty please?

promise to convert the next tests to criterion I am touching.

Comment (lbudai) [inline lib/str-repr/tests/test_decode.c]: fine :)

File: lib/str-repr/decode.c (hunk 2 above)
Comment (lbudai) [inline lib/str-repr/decode.c]: Yep.

File: lib/str-repr/encode.c (hunk 1 above)
Comment (lbudai) [inline lib/str-repr/encode.c]: `\" is` is used by default somewhere else, and using the same everywhere would be more symmetric

File: lib/str-repr/decode.c (hunk 2 above)
Comment (bazsi) [inline lib/str-repr/decode.c]: I have just addressed this in a subsequent patch. Note that I've refactored the function in about 6-8 steps, from which I had 6 separate commits, but I've folded them up, I hope it's ok.

Tests didn't change.

File: lib/str-repr/encode.c (hunk 1 above)
Comment (bazsi) [inline lib/str-repr/encode.c]: I don't see where else you saw this as default. I've looked for it, but no match.

It is true that we prefer double quotes if for some reason something is needed. Hey, I've changed it, easier than another iteration.

**Comment by bazsi** [comment]
Apart from criterion, I've submitted changes to all open issues.

Please merge if those are ok.

File: lib/str-repr/decode.c [hunk 5]
```diff
@@ -0,0 +1,212 @@
...
+      cur++;
+    }
+finish:
+  *end = cur;
+  /* check if quotation was not finished or we had extra characters, return FALSE */
+  if (quote_state == KV_QUOTE_INITIAL ||
+      quote_state == KV_EXPECT_DELIMITER ||
+      quote_state == KV_UNQUOTED_CHARACTERS)
```
Comment (bazsi) [inline lib/str-repr/decode.c]: Oops, i did not address this, however this fragment is now in a much smaller function, but i am not at my computer.

**Comment by bazsi** [comment]
The last patch should extract the complicated if condition into a separate function.

So with this, really all notes have been addressed.

**Comment by lbudai** [comment]
@bazsi : thanks!

**Pull request merged**